frontends/p4/strengthReduction.cpp

```diff
@@ -271,12 +271,22 @@ const IR::Node* DoStrengthReduction::postorder(IR::Slice* expr) {
             shift_amt = -k->asInt();
             shift_of = sh->left; } }
     if (shift_of) {
+        if (!shift_of->type->is<IR::Type_Bits>())
+            return expr;
+        if (shift_of->type->to<IR::Type_Bits>()->isSigned)
+            return expr;
         int hi = expr->getH();
         int lo = expr->getL();
         if (lo + shift_amt >= 0 && hi + shift_amt < shift_of->type->width_bits()) {
             expr->e0 = shift_of;
             expr->e1 = new IR::Constant(hi + shift_amt);
-            expr->e2 = new IR::Constant(lo + shift_amt); } }
+            expr->e2 = new IR::Constant(lo + shift_amt); }
+        if (lo + shift_amt < 0) {
+            expr->e0 = shift_of;
+            expr->e1 = new IR::Constant(hi + shift_amt);
+            expr->e2 = new IR::Constant(0);
```

Comment:
maybe should call `expr = postorder(expr);` here to recursively deal with anything nested in the expresion?  Or perhaps not worth it for such an obscure case.